orthoinference/src/main/java/org/reactome/orthoinference/StableIdentifierGenerator.java

```diff
@@ -1,67 +1,73 @@
+    private MySQLAdaptor dba;
+    private String speciesAbbreviation;
+
+    public StableIdentifierGenerator(MySQLAdaptor dba, String speciesAbbreviation) {
+        this.dba = dba;
+        this.speciesAbbreviation = speciesAbbreviation;
+    }
+
+    public GKInstance generateOrthologousStableId(GKInstance inferredInst, GKInstance originalInst) throws Exception {
 
         // All Human PhysicalEntitys and Events will have a StableIdentifier instance in the stableIdentifier attribute
         GKInstance stableIdentifierInst = (GKInstance) originalInst.getAttributeValue(stableIdentifier);
+        if (stableIdentifierInst == null) {
+            logger.fatal("No stable identifier instance found for " + originalInst + " -- terminating");
+            throw new RuntimeException("Could not find stable identifier instance");
+        }
 
         // For now, Human is hard-coded as the source species, so we replace the stableIdentifier source species based on that assumption
         String sourceIdentifier = (String) stableIdentifierInst.getAttributeValue(identifier);
         String targetIdentifier = sourceIdentifier.replace("HSA", speciesAbbreviation);
+
         // Paralogs will have the same base stable identifier, but we want to denote when that happens.
         // We pull the value from `seenOrthoIds`, increment it and then add it to the stable identifier name (eg: R-MMU-123456-2)
-        if (seenOrthoIds.get(targetIdentifier) == null) {
-            seenOrthoIds.put(targetIdentifier, 1);
-        } else {
-            int paralogCount = seenOrthoIds.get(targetIdentifier);
-            paralogCount++;
-            seenOrthoIds.put(targetIdentifier, paralogCount);
+        int paralogCount = Optional.ofNullable(seenOrthoIds.get(targetIdentifier)).orElse(0) + 1;
+        if (paralogCount > 1) {
             targetIdentifier += "-" + paralogCount;
         }
+        seenOrthoIds.put(targetIdentifier, paralogCount);
 
-        // Create new StableIdentifier instance
-        GKInstance orthoStableIdentifierInst = InstanceUtilities.createNewInferredGKInstance(stableIdentifierInst);
-        orthoStableIdentifierInst.addAttributeValue(identifier, targetIdentifier);
-        String identifierVersionNumber = "1";
-        orthoStableIdentifierInst.addAttributeValue(identifierVersion, identifierVersionNumber);
-        String orthoStableIdentifierName = targetIdentifier + "." + identifierVersionNumber;
-        orthoStableIdentifierInst.setDisplayName(orthoStableIdentifierName);
-        dba.storeInstance(orthoStableIdentifierInst);
+        // Check that the stable identifier instance does not already exist in DB
+        // TODO: Performance check
+        Collection<GKInstance> existingStableIdentifier = (Collection<GKInstance>) dba.fetchInstanceByAttribute("StableIdentifier", "identifier", "=", targetIdentifier);
+
+        GKInstance orthoStableIdentifierInst;
+        if (existingStableIdentifier.size() == 0) {
```

Comment:
Could the logic in this if block be a method for creating the stable id?

Reply:
Done